File: leetcode/704.BinarySearch.go
package leetcode

import "math"

func search(nums []int, target int) int {
	start := 0
	end := len(nums) - 1

	for start <= end {
		middleIndex := int(math.Floor(float64((start + end) / 2)))
		current := nums[middleIndex]

		if target == current {
			return middleIndex
		}

		if target < current {
			end = middleIndex - 1
		} else {
			start = middleIndex + 1
		}
	}

	return -1
}

File: leetcode/704.BinarySearch.ts
function search(nums: number[], target: number): number {
  let left = 0
  let right = nums.length - 1

  while (left <= right) {
    const middle = Math.floor((left + right) / 2)
    const value = nums[middle]

    if (value === target) return middle
    if (value < target) {
      left = left + 1
    } else {
      right = right - 1
    }
  }

  return -1
}
console.log(search([-1, 0, 3, 5, 9, 12], 9))
